### ISSUE TITLE
Linkerd is not classifying GRPC requests based on grpc-status headers.

### ISSUE THREAD

**Issue by abramoy** [linkerd/linkerd2#11947]
### What is the issue?

Hello,

GRPC requests are not being properly classified based on `grpc-status` header and hence all requests are marked as `HTTP_OK`.

To validate this I sent GRPC requests to a service which responds with a valid `grpc-status` header. Using Wireshark I am able to capture the response containing the header.
<img width="147" alt="Screenshot 2024-01-18 at 14 55 59" src="https://github.com/linkerd/linkerd2/assets/25409535/4aa9d099-c681-46c2-8388-b97d803b8e77">

Port forwarding to the proxy pod and querying the metrics API also returns unprocessed `grpc_status` attribute. **Note** `grpc_status=""` value.

```
response_total{direction="inbound",tls="no_identity",no_tls_reason="no_tls_from_remote",srv_group="",srv_kind="default",srv_name="all-unauthenticated",route_group="",route_kind="default",route_name="probe",authz_group="",authz_kind="default",authz_name="probe",status_code="200",classification="success",grpc_status="",error=""} 9013
```

### How can it be reproduced?

To reproduce:

- Enable Prometheus metric
- Send a GRPC request and respond with a valid `grpc-status` code. Image above used 14.
- Query the metrics api in the proxy
- Validate missing classification on `grpc_status` attribute

### Logs, error output, etc

```
response_total{direction="inbound",tls="no_identity",no_tls_reason="no_tls_from_remote",srv_group="",srv_kind="default",srv_name="all-unauthenticated",route_group="",route_kind="default",route_name="probe",authz_group="",authz_kind="default",authz_name="probe",status_code="200",classification="success",grpc_status="",error=""} 9013
```

### output of `linkerd check -o short`

```
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ control plane pods are ready
√ cluster networks contains all pods
√ cluster networks contains all services

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ proxy-init container runs as root user if docker container runtime is used

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
‼ issuer cert is valid for at least 60 days
    issuer certificate will expire on 2023-12-28T07:32:15Z
    see https://linkerd.io/2.14/checks/#l5d-identity-issuer-cert-not-expiring-soon for hints
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
‼ proxy-injector cert is valid for at least 60 days
    certificate will expire on 2024-01-01T13:43:33Z
    see https://linkerd.io/2.14/checks/#l5d-proxy-injector-webhook-cert-not-expiring-soon for hints
√ sp-validator webhook has valid cert
‼ sp-validator cert is valid for at least 60 days
    certificate will expire on 2024-01-04T08:56:13Z
    see https://linkerd.io/2.14/checks/#l5d-sp-validator-webhook-cert-not-expiring-soon for hints
√ policy-validator webhook has valid cert
‼ policy-validator cert is valid for at least 60 days
    certificate will expire on 2023-12-28T09:31:39Z
    see https://linkerd.io/2.14/checks/#l5d-policy-validator-webhook-cert-not-expiring-soon for hints

linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    is running version 2.14.3 but the latest stable version is 2.14.6
    see https://linkerd.io/2.14/checks/#l5d-version-cli for hints

control-plane-version
---------------------
√ can retrieve the control plane version
‼ control plane is up-to-date
    is running version 2.14.3 but the latest stable version is 2.14.6
    see https://linkerd.io/2.14/checks/#l5d-version-control for hints
√ control plane and cli versions match

linkerd-control-plane-proxy
---------------------------
√ control plane proxies are healthy
‼ control plane proxies are up-to-date
    some proxies are not running the current version:
	* linkerd-destination-7565484496-67p6h (v2.210.2)
	* linkerd-destination-7565484496-6bm7j (v2.210.2)
	* linkerd-destination-7565484496-f5lbf (v2.210.2)
	* linkerd-destination-7565484496-kfbpc (v2.210.2)
	* linkerd-destination-7565484496-ld969 (v2.210.2)
	* linkerd-identity-78fb66464f-5ptjq (v2.210.2)
	* linkerd-identity-78fb66464f-6h256 (v2.210.2)
	* linkerd-identity-78fb66464f-9qjvf (v2.210.2)
	* linkerd-identity-78fb66464f-t5h6w (v2.210.2)
	* linkerd-identity-78fb66464f-vbrwm (v2.210.2)
	* linkerd-proxy-injector-5c55846f8c-26ksp (v2.210.2)
	* linkerd-proxy-injector-5c55846f8c-7q4ch (v2.210.2)
	* linkerd-proxy-injector-5c55846f8c-9h9mc (v2.210.2)
	* linkerd-proxy-injector-5c55846f8c-qmhkq (v2.210.2)
	* linkerd-proxy-injector-5c55846f8c-sfwfd (v2.210.2)
    see https://linkerd.io/2.14/checks/#l5d-cp-proxy-version for hints
‼ control plane proxies and cli versions match
    linkerd-destination-7565484496-67p6h running v2.210.2 but cli running stable-2.14.3
    see https://linkerd.io/2.14/checks/#l5d-cp-proxy-cli-version for hints

linkerd-ha-checks
-----------------
√ pod injection disabled on kube-system
√ multiple replicas of control plane pods

linkerd-viz
-----------
√ linkerd-viz Namespace exists
√ can initialize the client
√ linkerd-viz ClusterRoles exist
√ linkerd-viz ClusterRoleBindings exist
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ tap API service is running
‼ linkerd-viz pods are injected
    could not find proxy container for prometheus-scrape-1-5585795fbd-fpcxq pod
    see https://linkerd.io/2.14/checks/#l5d-viz-pods-injection for hints
‼ viz extension pods are running
    container "linkerd-proxy" in pod "prometheus-scrape-1-5585795fbd-fpcxq" is not ready
    see https://linkerd.io/2.14/checks/#l5d-viz-pods-running for hints
√ viz extension proxies are healthy
‼ viz extension proxies are up-to-date
    some proxies are not running the current version:
	* grafana-76f7b4b7f9-9pbg4 (v2.210.2)
	* metrics-api-dcb7f4875-jjzj4 (v2.210.2)
	* metrics-api-dcb7f4875-xbkg6 (v2.210.2)
	* tap-7dfcf4686-76m9p (v2.210.2)
	* tap-injector-699b466bbd-whdzr (v2.210.2)
	* web-59b65979b5-5hjxg (v2.210.2)
    see https://linkerd.io/2.14/checks/#l5d-viz-proxy-cp-version for hints
‼ viz extension proxies and cli versions match
    grafana-76f7b4b7f9-9pbg4 running v2.210.2 but cli running stable-2.14.3
    see https://linkerd.io/2.14/checks/#l5d-viz-proxy-cli-version for hints
√ prometheus is installed and configured correctly
√ viz extension self-check

linkerd-smi
-----------
√ linkerd-smi extension Namespace exists
√ SMI extension service account exists
√ SMI extension pods are injected
√ SMI extension pods are running
√ SMI extension proxies are healthy

Status check results are √
```

### Environment

- EKS
- Kubernetes 1.25
- Linkerd 2.14.3
- Service
  - Spring Boot 3
  - [net.devh](https://mvnrepository.com/artifact/net.devh) » `grpc-server-spring-boot-starter 2.15.0.RELEASE`

### Possible solution

We looked at the classify class. It looks like the logic is built but this does not seem to work.
https://github.com/linkerd/linkerd2-proxy/blob/main/linkerd/app/core/src/classify.rs#L326

### Additional context

I enabled both trace logs in the proxy and the debug container in pod. This did not help find the root cause.

### Would you like to work on fixing this bug?

maybe

**Comment by adleong** [comment]
hi @abramoy,

I see that you're using stable-2.14.3.  An issue about missing grpc_status labels that matches your description was fixed in stable-2.14.4 (https://github.com/linkerd/linkerd2/releases/tag/stable-2.14.4).  I recommend you try the latest stable release and see if that solves the issue for you.

**Comment by abramoy** [comment]
Hi @adleong, thanks very much for the reply. I must have missed the release log. That is now fixed.
